src/librbd/DeepCopyRequest.cc

```diff
@@ -215,13 +215,18 @@ void DeepCopyRequest<I>::send_copy_object_map() {
 
   // rollback the object map (copy snapshot object map to HEAD)
   RWLock::WLocker object_map_locker(m_dst_image_ctx->object_map_lock);
-  auto ctx = new FunctionContext([this, finish_op_ctx](int r) {
-      handle_copy_object_map(r);
-      finish_op_ctx->complete(0);
-    });
   ceph_assert(m_snap_seqs->count(m_snap_id_end) > 0);
   librados::snap_t copy_snap_id = (*m_snap_seqs)[m_snap_id_end];
+
+  using klass = DeepCopyRequest<I>;
```

Comment:
Nit: in general, I think it would be preferable to hide the reference counting to within the affected classes. For example, in ```ObjectMap::rollback```, it could wrap the provided ```Context``` with a helper in ```Utils.h``` that invokes the ```get``` on construction and ```put``` on destruction. That should prevent most of this churn throughout librbd.

Reply:
Ack. Perhaps a wraparound like this? - https://gist.github.com/MahatiC/95fc2133127c17d4389d20511c2c3f76 

I think this would simplify the code quite a bit, though it wouldn't be applicable to all of the changes.

Reply:
Yeah -- I think something like that looks good. Perhaps rename the helper/wrapper to something including ref-counting (i.e. ```create_context_callback``` -> ```create_ref_counted_context```)